Unified CLI

pgierz opened the pull request:
Cleans up the CLI for a more unified experience:

## Feature List
* [x] All programs start with `esm-tools <command>`
* [x] Beautiful colours
* [x] Old commands still work as expected [![esm-tools main test](https://github.com/esm-tools/esm_tools/actions/workflows/esm_tools_actions.yml/badge.svg?branch=feat%2Funified_cli)](https://github.com/esm-tools/esm_tools/actions/workflows/esm_tools_actions.yml)

## To consider 
* https://github.com/esm-tools/esm_tools/labels/question Version and model are combined by minus. Maybe adopt a spack-like syntax?

## Test
* [x] `esm-tools get fesom-2.5` (Note: no added minus, the subcommand is *just* `get`)
* [x] `esm-tools comp fesom-2.5`
* [x] `esm-tools install fesom-2.5`
* [x] `esm-tools run ...`

pgierz committed:
--wip-- [skip ci]

pgierz committed:
merge...?

pgierz committed:
wip

pgierz committed:
style: proper class names

pgierz committed:
feat: translate esm_master cli to use click

pgierz committed:
wip: rework of esm-master

pgierz committed:
wip: archive

pgierz commented:
Incomplete, but here is a preview:

<img width="1459" alt="image" src="https://github.com/user-attachments/assets/7cee6a68-7a79-4a43-9faf-e832c0ec13da">

pgierz committed:
wip: esm-master still

pgierz committed:
wip: translates plugin-manager cli to click

pgierz committed:
wip: esm-database to click

pgierz committed:
wip: esm-tests to click

pgierz committed:
still working on esm-master

pgierz committed:
ci: new ci

pgierz committed:
wrong command name

pgierz commented:
This PR includes https://github.com/esm-tools/esm_tools/labels/continuous%20integration label, since there is a new CI step specifically for the new syntax, e.g.:

```console
$ esm-tools get <model>
```

[![esm-tools new cli test](https://github.com/esm-tools/esm_tools/actions/workflows/new-cli.yml/badge.svg)](https://github.com/esm-tools/esm_tools/actions/workflows/new-cli.yml)

pgierz committed:
wip

pgierz committed:
wip wip wip

pgierz committed:
ci: tests for new cli only cover get, for now

ackerlar committed:
add new awiesm version awiesm-2.6 for fesom-2.6

ackerlar committed:
update awiesm pism coupling scripts

pgierz committed:
wip: translates runscripts to rich click

pgierz commented:
![02_50_57](https://github.com/user-attachments/assets/fd09b4bc-a0d7-4387-991e-91a020d9f9ff)

Almost done, just need meta-commands inside esm-master.

pgierz committed:
wip: esm-master meta commands

pgierz committed:
Merge branch 'release' into feat/unified_cli

pgierz commented:
# Copilot Summary
This pull request includes several key changes to the `esm-tools` project, focusing on adding new CLI tests, updating dependencies, and refactoring CLI scripts to use `rich-click` for improved command-line interfaces.

### CI/CD Improvements:
* Added a new GitHub Actions workflow for testing the new CLI (`.github/workflows/new-cli.yml`).

### Dependency Updates:
* Updated `setup.py` to include new dependencies: `rich-click` and `click-repl` and restructured the `console_scripts` entry points. [[1]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R15-R16) [[2]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R51-R65) [[3]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L74-R90)

### CLI Refactoring:
* Refactored multiple CLI scripts to use `rich-click` for enhanced command-line interfaces:
  - `src/esm_archiving/cli.py`
  - `src/esm_cleanup/cli.py`
  - `src/esm_database/cli.py`
  - `src/esm_master/cli.py`
  - `src/esm_master/cli-broken.py`

### Code Clean-Up:
* Removed unused imports from `setup.py` and other minor code clean-ups. [[1]](diffhunk://#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7L5) [[2]](diffhunk://#diff-bc8244fdda32ea7d151b0b8792316af062d251700098583aca6fa0d012160689R3)

These changes enhance the project's maintainability and user experience by improving the CLI's usability and ensuring the CI/CD pipeline tests the new CLI functionality.

JanStreffing commented:
I don't see esm-master clean and recomp there. Has something changed in this regard?

Edit: In the second picture there is clean, but no recomp or install, which it seems are on your radar.

pgierz commented:
That is in there now, here is an up-to-date version:


![image](https://github.com/user-attachments/assets/2880a396-e5c8-46df-8678-c99e24f0119b)

Feel free to try it out and give feedback.

**NOTE** You need to install with `pip install .` (*not* in editable mode, I still need to fix that)

pgierz committed:
fix: meta-commands of esm-master

ackerlar committed:
change to fesom-2.6.1

ackerlar committed:
change type to bool form int in flags for pism coupling

ackerlar committed:
change logical flags in coupling to be consistent with BoolWithProvenance type

ackerlar committed:
allow monthly debm input in coupling

Commit:
Bump version: 6.42.2 → 6.43.0

JanStreffing committed:
Update oasis.py

Quote from oasis source code:

                 WRITE(nulprt1,*) '        ***WARNING*** load balancing analysis : '
                 WRITE(nulprt1,*) '        Since v5.0, a third parameter is required '
                 WRITE(nulprt1,*) '        Default value 0 used for time statistic and land balancing analysis '

Commit:
Bump version: 6.43.0 → 6.43.1

pgierz commented:
Todo: symlinks for configs, namelists, etc

pgierz committed:
symlink configs

pgierz committed:
symlink namelists

pgierz committed:
symlink runscripts

pgierz commented:
@mandresm I am not sure if the `setup.py` still needs to be adapted to get the right paths. You worked on that more recently than I did, maybe you can have a look?

seb-wahl commented:
Something that you/we should definitely check if the updated works correctly from an existing installation i.e. that the move of namelists folder and the symlinking afterwards works as expected.
